File: Sprint8/J/C++/main.cpp
// https://contest.yandex.ru/contest/26131/problems/J/
#include <iostream>
#include <vector>
#include <map>
#include <algorithm>
#include <sstream>

struct TrieNode {
    std::map<char, TrieNode*> children;
    std::vector<std::string> words;
};

TrieNode* createNode() {
    return new TrieNode();
}

void insert(TrieNode* root, const std::string &word) {
    TrieNode* node = root;
    for (char ch : word) {
        if (isupper(ch)) {
            if (!node->children.count(ch)) {
                node->children[ch] = createNode();
            }
            node = node->children[ch];
        }
    }

    node->words.push_back(word);
}

void build_search_result(TrieNode* current, std::vector<std::string>& result) {
    result.insert(result.begin(), current->words.begin(), current->words.end());
    for (auto entry : current->children) {
        build_search_result(entry.second, result);
    }
}

std::vector<std::vector<std::string>> search(TrieNode* root, std::vector<std::string> &requests) {
    std::vector<std::vector<std::string>> results;

    for (const auto& request: requests) {
        std::vector<std::string> result;
        TrieNode* node = root;
        bool no_found = false;

        for (char ch : request) {
            if (!node->children.count(ch)) {
                no_found = true;
                break;
            }
            node = node->children[ch];
        }
        if (no_found) {
            result.emplace_back("");
        } else {
            build_search_result(node, result);
            std::sort(result.begin(), result.end());
        }
        results.push_back(result);
    }

    return results;
}

int main() {
    int n;
    std::cin >> n;

    TrieNode* root = createNode();
    for (int i = 0; i < n; ++i) {
        std::string word;
        std::cin >> word;
        insert(root, word);
    }

    int m;
    std::cin >> m;
    std::cin.ignore();
    std::vector<std::string> requests(m);
    for (int i = 0; i < m; ++i) {
        std::string pattern;
        std::getline(std::cin, pattern);
        requests[i] = pattern;
    }

    std::vector<std::vector<std::string>> results = search(root, requests);
    for (const auto& result: results) {
        if (result.empty()) {
            std::cout << std::endl;
            continue;
        }

        for (const std::string &word : result) {
            std::cout << word << std::endl;
        }
    }

    return 0;
}

File: Sprint8/J/Rust/src/main.rs
// https://contest.yandex.ru/contest/26131/problems/J/
use std::collections::HashMap;
use std::io::BufRead;

struct TrieNode {
    children: HashMap<char, Box<TrieNode>>,
    words: Vec<String>,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: HashMap::new(),
            words: Vec::new(),
        }
    }

    fn insert(&mut self, word: &str) {
        let mut node = self;
        for ch in word.chars() {
            if ch.is_uppercase() {
                node = node
                    .children
                    .entry(ch)
                    .or_insert_with(|| Box::new(TrieNode::new()));
            }
        }
        node.words.push(word.to_string());
    }

    pub fn search(&self, request: &str) -> Vec<String> {
        let mut node = self;
        let mut result = Vec::new();

        for ch in request.chars() {
            if let Some(child) = node.children.get(&ch) {
                node = child;
            } else {
                return vec!["".to_string()];
            }
        }

        node.build_search_result(&mut result);
        result.sort_unstable();
        result
    }

    fn build_search_result(&self, result: &mut Vec<String>) {
        result.extend_from_slice(&self.words);
        for child in self.children.values() {
            child.build_search_result(result);
        }
    }
}

fn main() {
    let mut lines = std::io::stdin().lock().lines();

    let n: usize = lines.next().unwrap().unwrap().parse().unwrap();
    let mut root = TrieNode::new();
    for _ in 0..n {
        let word = lines.next().unwrap().unwrap();
        root.insert(&word);
    }

    let m: usize = lines.next().unwrap().unwrap().parse().unwrap();
    let requests: Vec<String> = (0..m).map(|_| lines.next().unwrap().unwrap()).collect();

    for request in &requests {
        let results = root.search(request);
        for word in results {
            println!("{}", word);
        }
    }
}
